src/Functional/Curry.php

```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * Copyright (C) 2011-2016 by Gilles Crettenand <gilles@crettenand.info>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+namespace Functional;
+
+use ReflectionMethod;
+use ReflectionFunction;
+
+/**
+ * Return a curryied version of the given function. You can decide if you also
+ * want to curry optional parameters or not.
+ *
+ * @param callable $function the function to curry
+ * @param bool $required curry optional parameters ?
+ * @return callable a curryied version of the given function
+ */
+function curry(callable $function, $required = true)
+{
+    if (is_string($function) && strpos($function, '::', 1) !== false) {
```

Comment:
We will still have to use reflection to get the number of parameters tho. According to the RFC there will however be some performance gains.

What do you want me to do ? detect if the method exists and use it if it is the case ?